torch_frame/nn/encoder/stype_encoder.py

```diff
@@ -744,3 +783,67 @@ def encode_forward(
         # -> [batch_size, num_cols, out_channels]
         x = x_lin + self.bias
         return x
+
+
+class TimestampEncoder(StypeEncoder):
+    r"""TimestampEncoder for timestamp stype. The encoder applies
+    different frequencies to the timestamps to calculate the
+    positional encoding and then adds the positional encoding to
+    the original :obj:`Tensor`. Year is encoded with
+    :class:`torch_frame.nn.encoding.PositionalEncoding`. The other
+    features, including month, day, dayofweek, hour, minute and second,
+    are encoded using :class:`torch_frame.nn.encoding.CyclicEncoding`.
+    We then use a :class:`~torch.nn.Linear` layer to map back to
+    out_channels.
+
+    Args:
+        out_size (int): Output dimension of the positional and cyclic
+            encoding.
+    """
+    supported_stypes = {stype.timestamp}
+
+    def __init__(
+        self,
+        out_channels: Optional[int] = None,
+        stats_list: Optional[List[Dict[StatType, Any]]] = None,
+        stype: Optional[stype] = None,
+        post_module: Optional[Module] = None,
+        na_strategy: Optional[NAStrategy] = None,
+        out_size: int = 8,
+    ):
+        self.out_size = out_size
+        super().__init__(out_channels, stats_list, stype, post_module,
+                         na_strategy)
+
+    def init_modules(self):
+        super().init_modules()
+        min_year = torch.tensor([
+            self.stats_list[i][StatType.YEAR_RANGE][0]
+            for i in range(len(self.stats_list))
+        ])
+        self.register_buffer("min_year", min_year)
+        self.positional_encoding = PositionalEncoding(self.out_size)
+        self.cyclic_encoding = CyclicEncoding(self.out_size)
+        self.linear = Linear(
+            len(TIME_TO_INDEX) * self.out_size, self.out_channels)
+        self.reset_parameters()
+
+    def reset_parameters(self):
+        super().reset_parameters()
+        self.linear.reset_parameters()
+
+    def encode_forward(self, feat: Tensor,
+                       col_names: Optional[List[str]] = None) -> Tensor:
+        feat = feat.to(torch.float32)
+        max_values = torch.tensor([
+            1, NUM_MONTHS_PER_YEAR, NUM_DAYS_PER_MONTH, NUM_DAYS_PER_WEEK,
+            NUM_HOURS_PER_DAY, NUM_MINUTES_PER_HOUR, NUM_SECONDS_PER_MINUTE
+        ])
+        feat[..., TIME_TO_INDEX['YEAR']] = feat[:, :, 0] - self.min_year
+        feat /= max_values.view(1, -1)
+        encodings = [
+            self.positional_encoding(feat[..., :1]),
+            self.cyclic_encoding(feat[..., 1:])
+        ]
+        out = torch.cat(encodings, dim=2).view(feat.size(0), feat.size(1), -1)
```

Comment:
nit: avoid using "-1", since it will error out when feat is emty tensor (can happen when neighbor smapling does not sample any node for this type)